core/src/main/java/io/grpc/internal/AbstractClientStream.java

```diff
@@ -218,6 +235,19 @@ protected final ClientStreamListener listener() {
     protected void inboundHeadersReceived(Metadata headers) {
       Preconditions.checkState(!statusReported, "Received headers on closed stream");
       statsTraceCtx.clientInboundHeaders();
+
+      Decompressor decompressor = Codec.Identity.NONE;
+      if (headers.containsKey(MESSAGE_ENCODING_KEY)) {
+        String encoding = headers.get(MESSAGE_ENCODING_KEY);
+        decompressor = decompressorRegistry.lookupDecompressor(encoding);
+        if (decompressor == null) {
+          deframeFailed(Status.INTERNAL.withDescription(
+              String.format("Can't find decompressor for %s", encoding)).asException());
```

Comment:
`asRuntimeException()` produces identical behavior. The exception is thrown away and only the status is kept in all `deframeFailed()` handling. And `deframeFailed()` is also the only way the `TransportState` can kill the RPC.